Modules/Core/Common/include/itkMacro.h

```diff
@@ -316,31 +316,24 @@ namespace itk
 #define itkSimpleNewMacro(x)                                                                                           \
   static Pointer New()                                                                                                 \
   {                                                                                                                    \
-    Pointer smartPtr = ::itk::ObjectFactory<x>::Create();                                                              \
-    if (smartPtr == nullptr)                                                                                           \
+    Pointer smartPtr{ ::itk::ObjectFactory<x>::Create() };                                                             \
+    if (smartPtr.IsNull())                                                                                             \
     {                                                                                                                  \
       smartPtr = new x;                                                                                                \
     }                                                                                                                  \
-    smartPtr->UnRegister();                                                                                            \
     return smartPtr;                                                                                                   \
   }                                                                                                                    \
   ITK_MACROEND_NOOP_STATEMENT
 
 #define itkCreateAnotherMacro(x)                                                                                       \
   ::itk::LightObject::Pointer CreateAnother() const override                                                           \
   {                                                                                                                    \
-    ::itk::LightObject::Pointer smartPtr;                                                                              \
-    smartPtr = x::New().GetPointer();                                                                                  \
-    return smartPtr;                                                                                                   \
+    return ::itk::LightObject::Pointer{ x::New().GetPointer() };                                                       \
```

Comment:
Is there a move constructor which could be used here?

Reply:
@blowekamp Thank you for the excellent feedback.

I have not tried Valgrind.  I will figure out how that is done with ITK ... though if you have a quick pointer on how to do that, that would be much appreciated.

There could be an exception someplace that I missed, but I believe the extra reference count is not exposed through the public API.  For example, it is just before the `SmartPointer` is exposed when the extra reference count (whether from `Create` or `new`) is UnRegistered in the original code for `x::New()`: https://github.com/InsightSoftwareConsortium/ITK/blob/eec9fe67056bfb5c693a4915b86eac9271e1dc83/Modules/Core/Common/include/itkMacro.h#L316-L327

(Or am I being too dense to understand that you are saying that the public API of ITK includes `new x` and/or `ObjectFactory<x>::Create()`?)

> ITK objects can be owned by other objects besides smart pointers.

Please tell me more!

The code change doesn't change the public API (that I know) and doesn't fix any existing bug (that I know) but is primarily to head off future bugs. With the original implementation, someone who uses `new x` or `ObjectFactory<x>::Create()` inside of ITK is often required to also invoke `UnRegister()`, and whether that is needed or not depends upon a good understanding of our non-standard implementation.
 
> Perhaps at the next major ITK release the objects interface for smart pointers could be refined but I am opposed to it during minor release changes.

For me, it hinges on whether there are any public API changes, such as the uses via other than `SmartPointer` that you mention.

Reply:
Very good.  I just pushed a commit that uses the move constructor in situations like this.  The one place that I did not make the change involved an explicit `dynamic_cast`, which could return `nullptr`.

Reply:
I just pushed a commit that undoes the one just mentioned.  In an effort to keep things simple, the latest commit does pretty much the bare minimum to adjust reference counts to (what I consider to be) the accepted standard.  I will put more details about that in the general conversation comments in just a moment.